Assets/MixedRealityToolkit/_Core/Sharing/AppSystems/Sessions/SessionStageBase.cs

```diff
@@ -0,0 +1,179 @@
+﻿using Pixie.StateControl;
+using System.Collections;
+using UnityEngine;
+using Pixie.AppSystems.TimeSync;
+
+namespace Pixie.AppSystems.Sessions
+{
+    public abstract class SessionStageBase : MonoBehaviour, ISessionStage
+    {
+        public float MaxTime { get { return maxTime; } }
+
+        public float TimeStarted { get { return timeStarted; } }
+
+        public float TimeElapsed
+        {
+            get
+            {
+                switch (state)
+                {
+                    case StageStateEnum.NotStarted:
+                        return 0f;
+
+                    case StageStateEnum.Completed:
+                        return timeElapsed;
+
+                    default:
+                        return timeSource.Time - timeStarted;
+                }
+            }
+        }
+
+        public StageProgressionTypeEnum ProgressionType { get { return type; } }
+        public StageStateEnum State { get { return state; } }
+
+        [Header("General Settings")]
+        [SerializeField]
+        private StageProgressionTypeEnum type = StageProgressionTypeEnum.Timed;
+        [SerializeField]
+        protected StageStateEnum state = StageStateEnum.NotStarted;
+        [SerializeField]
+        protected float maxTime = 600f;
+
+        private float timeStarted;
+        private float timeElapsed;
+
+        private ITimeSource timeSource;
+        private IAppStateReadWrite appState;
+
+        private bool forceMoveNext = false;
+
+        public void ForceComplete()
+        {
+            forceMoveNext = true;
+        }
+
+        public abstract void OnStageEnter(ITimeSource timeSource, IAppStateReadWrite appState);
```

Comment:
What's the expected behavior for these OnStageEnter and OnStageExit calls? Above, they aren't implemented. Is the intent of these methods fulfilled with TransitionIn & TransitionOut?

Reply:
OnStageEnter / Exit are intended for instantaneous transitions. Eg in an arcade shooter this might switch the behavior mode of all the enemies in the level.

TransitionIn and TransitionOut are intended for cases where you either want a smooth transition from one state to another, or when you don't want the stage to actually begin until some condition is met. Eg, waiting for all joined users to set their state to 'Ready' before a meeting starts.

Reply:
can we get an overview/timeline for all of these events added to the description?